vectorstores/inmemory/inmemory.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	ids := make([]string, len(vectors))
+	for i, vec := range vectors {
+		s.Lock()
+
+		id := s.lastID + 1
+		if id > uint32(s.sizeLimit) {
+			s.index.Grow(2 * s.sizeLimit)
+			s.sizeLimit *= 2
+		}
+		s.index.Add(vec, id)
+		s.lastID = id
+
+		s.content[id] = docs[i].PageContent
+		s.meta[id] = docs[i].Metadata
+
+		s.Unlock()
+
+		ids[i] = strconv.FormatUint(uint64(id), 10)
+	}
+
+	return ids, nil
+}
+
+func (s *Store) SimilaritySearch(
+	ctx context.Context,
+	query string,
+	numDocuments int,
+	options ...vectorstores.Option,
+) ([]schema.Document, error) {
+	opts := s.getOptions(options...)
+	if opts.NameSpace != "" {
+		// in-memory store does not support these options
+		return nil, ErrUnsupportedOptions
+	}
+
+	var filters map[string]any
+	if f, ok := opts.Filters.(map[string]any); ok {
+		filters = f
+	}
+
+	embedder := s.embedder
+	if opts.Embedder != nil {
+		embedder = opts.Embedder
+	}
+	embedderData, err := embedder.EmbedQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	docs := make([]schema.Document, 0, numDocuments)
+	result := s.index.Search(embedderData, s.efSearch, numDocuments)
+	for {
+		item := result.Pop()
+		if item == nil {
+			break
+		}
+
+		// skip ID 0 which is the first point
+		if item.ID == 0 {
+			continue
+		}
+
+		s.RLock()
+
+		id := item.ID
+		doc := schema.Document{}
+		doc.PageContent = s.content[id]
+		doc.Metadata = s.meta[id]
+		doc.Score = 1.0 - item.D // convert distance to similarity
+
+		s.RUnlock()
+
+		docs = append(docs, doc)
+	}
+
+	docs = applyFilters(docs, filters)
+	docs, err = applyScoreThreshold(docs, opts.ScoreThreshold)
+	if err != nil {
+		return nil, err
+	}
+
+	// reverse the slice to get the highest score first
+	slices.Reverse(docs)
+	return docs, nil
+}
+
+// getOptions applies given options to default Options and returns it
+// This uses options pattern so clients can easily pass options without changing function signature.
+func (s *Store) getOptions(options ...vectorstores.Option) vectorstores.Options {
+	opts := vectorstores.Options{}
+	for _, opt := range options {
+		opt(&opts)
+	}
+	return opts
+}
+
+// deduplicate applies the deduplicater to the given slice of documents.
+// It returns a new slice of documents with the duplicates removed.
+func (s *Store) deduplicate(
+	ctx context.Context,
+	opts vectorstores.Options,
+	docs []schema.Document,
+) []schema.Document {
+	if opts.Deduplicater == nil {
+		return docs
+	}
+
+	filtered := make([]schema.Document, 0, len(docs))
+	for _, doc := range docs {
+		if !opts.Deduplicater(ctx, doc) {
+			filtered = append(filtered, doc)
+		}
+	}
+
+	return filtered
+}
+
+// cosineDistance returns the cosine distance between two vectors.
+func cosineDistance(x, y []float32) float32 {
+	if len(x) != len(y) {
+		panic("vectors must have the same length")
+	}
+
+	var dot, s1, s2 float32
+	for i := 0; i < len(x); i++ {
+		dot += x[i] * y[i]
+		s1 += math32.Pow(x[i], 2)
+		s2 += math32.Pow(y[i], 2)
```

Comment:
A simple multiplication is probably faster than a function call.

Reply:
thanks. i fix it